// jupiter-api-client/dist/index.js
// src/utils/fetch.ts
class JupiterAPIError extends Error {
  status;
  statusText;
  response;
  constructor(message, status, statusText, response) {
    super(message);
    this.status = status;
    this.statusText = statusText;
    this.response = response;
    this.name = "JupiterAPIError";
  }
}
async function fetchWithErrorHandling(url, config) {
  try {
    const urlObj = new URL(url);
    if (config?.params) {
      Object.entries(config.params).forEach(([key, value]) => {
        if (value !== undefined) {
          urlObj.searchParams.append(key, String(value));
        }
      });
    }
    let abortController;
    let timeoutId;
    if (config?.timeout !== undefined && config.timeout > 0 || config?.signal !== undefined) {
      abortController = new AbortController;
      if (config.signal !== undefined) {
        config.signal.addEventListener("abort", () => abortController?.abort());
      }
      if (config.timeout !== undefined && config.timeout > 0) {
        timeoutId = setTimeout(() => abortController?.abort(), config.timeout);
      }
    }
    const response = await fetch(urlObj.toString(), {
      method: "GET",
      headers: {
        Accept: "application/json",
        ...config?.headers
      },
      signal: abortController?.signal
    });
    if (timeoutId) {
      clearTimeout(timeoutId);
    }
    if (!response.ok) {
      let errorMessage = `HTTP ${response.status}: ${response.statusText}`;
      let errorData;
      try {
        errorData = await response.json();
        if (typeof errorData === "object" && errorData !== null && "message" in errorData) {
          const err = errorData;
          errorMessage = err.message;
        }
      } catch {}
      throw new JupiterAPIError(errorMessage, response.status, response.statusText, errorData);
    }
    const data = await response.json();
    return data;
  } catch (error) {
    if (error instanceof JupiterAPIError) {
      throw error;
    }
    if (error instanceof Error) {
      if (error.name === "AbortError") {
        throw new JupiterAPIError("Request timeout");
      }
      throw new JupiterAPIError(`Network error: ${error.message}`);
    }
    throw new JupiterAPIError("Unknown error occurred");
  }
}

// src/api/config.ts
var JUPITER_API_BASE_URL = "https://lite-api.jup.ag";
var ENDPOINTS = {
  TOKENS: {
    SEARCH: "/tokens/v2/search",
    TAG: "/tokens/v2/tag",
    CATEGORY: "/tokens/v2",
    RECENT: "/tokens/v2/recent"
  },
  PRICE: {
    GET: "/price/v3"
  }
};

// src/api/tokens.ts
class TokensAPI {
  baseUrl;
  constructor(baseUrl = JUPITER_API_BASE_URL) {
    this.baseUrl = baseUrl;
  }
  async search(options) {
    const url = new URL(`${this.baseUrl}${ENDPOINTS.TOKENS.SEARCH}`);
    url.searchParams.append("query", options.query);
    return fetchWithErrorHandling(url.toString());
  }
  async getByTag(options) {
    const url = new URL(`${this.baseUrl}${ENDPOINTS.TOKENS.TAG}`);
    url.searchParams.append("query", options.query);
    return fetchWithErrorHandling(url.toString());
  }
  async getByCategory(options) {
    const url = new URL(`${this.baseUrl}${ENDPOINTS.TOKENS.CATEGORY}/${options.category}/${options.interval}`);
    if (options.limit !== undefined) {
      url.searchParams.append("limit", options.limit.toString());
    }
    return fetchWithErrorHandling(url.toString());
  }
  async getRecent() {
    const url = `${this.baseUrl}${ENDPOINTS.TOKENS.RECENT}`;
    return fetchWithErrorHandling(url);
  }
}

// src/api/price.ts
class PriceAPI {
  baseUrl;
  constructor(baseUrl = JUPITER_API_BASE_URL) {
    this.baseUrl = baseUrl;
  }
  async get(options) {
    const url = new URL(`${this.baseUrl}${ENDPOINTS.PRICE.GET}`);
    const idsParam = Array.isArray(options.ids) ? options.ids.join(",") : options.ids;
    url.searchParams.append("ids", idsParam);
    return fetchWithErrorHandling(url.toString());
  }
  async getOne(mintAddress) {
    const response = await this.get({ ids: mintAddress });
    return response[mintAddress];
  }
  async getMany(mintAddresses) {
    return this.get({ ids: mintAddresses });
  }
}

// src/client.ts
class JupiterAPIClient {
  tokens;
  price;
  baseUrl;
  constructor(options = {}) {
    this.baseUrl = options.baseUrl ?? JUPITER_API_BASE_URL;
    this.tokens = new TokensAPI(this.baseUrl);
    this.price = new PriceAPI(this.baseUrl);
  }
}
// src/index.ts
var src_default = JupiterAPIClient;
export {
  src_default as default,
  TokensAPI,
  PriceAPI,
  JupiterAPIError,
  JupiterAPIClient
};
